### PR TITLE
Support maxDistance and minDistance in geo filter

### PR DESCRIPTION
**Pull request by Tuxity**
Hi,

This PR bring the support of maxDistance and minDistance for geoFilter 

minDistance is supported but will not be used until it's handle in https://github.com/strongloop/loopback-datasource-juggler

Cleaned up from https://github.com/strongloop/loopback-connector-mongodb/pull/259 with more fix.
Maybe the tests part is not very optimized, but it's working.

Don't hesitate to comment, I will update it to merge it asap

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by superkhau** [comment]
Looks like duplicate of https://github.com/strongloop/loopback-connector-mongodb/pull/259#issuecomment-241899854

**Comment by superkhau** [comment]
Actually I see you're just basing your work on #259. Please fix linting errors:

```
**************************************************
**
**  Linting commit logs
**
**  2 problems found:
**    1c4a5cb - Support maxDistance and minDistance in geo filter: Second line should be empty.<minDistance is supported but will not be used until it's handle in https://github.com/strongloop/loopback-datasource-juggler>
**    1c4a5cb - Support maxDistance and minDistance in geo filter: line 4 longer than 72 characters.
**
**************************************************
```

See https://github.com/strongloop/loopback-contributor-docs/blob/master/git-commit-messages.md for more info.

**Comment by Tuxity** [comment]
@superkhau It's done, lemme know if there is other thing to modify

File: test/mongodb.test.js [hunk 1]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
       });
     });
 
+    it('find should be able to query by location via near with maxDistance', function(done) {
+      var coords = { lat: 30.274085, lng: 120.15507000000002 };
+
+      geoDb.autoupdate(function(err) {
+        var createPosts = function(far) {
```
Comment (superkhau) [inline test/mongodb.test.js]: Please move this function outside of the test and share it with the other 3 tests.

```
// geopoint being {lat: i, lng: j}
function createPosts(isFar, geopoint) {
  ...
}
Then just pass in the differences.

File: test/mongodb.test.js [hunk 2]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+          };
+        };
+
+        async.parallel([
+          createPosts(false),
+          createPosts(false),
+          createPosts(false),
+          createPosts(true),
+        ], function(err) {
+          should.not.exist(err);
```
Comment (superkhau) [inline test/mongodb.test.js]: `if (err) return done(err);` for better stack traces.

File: test/mongodb.test.js [hunk 3]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+          PostWithLocation.find({
+            where: {
+              location: {
+                near: new GeoPoint(coords),
+                maxDistance: 17000,
+                unit: 'meters',
+              },
+            },
+          }, function(err, results) {
+            should.not.exist(err);
```
Comment (superkhau) [inline test/mongodb.test.js]: ditto

File: test/mongodb.test.js [hunk 4]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+            where: {
+              location: {
+                near: new GeoPoint(coords),
+                minDistance: 17000,
+                unit: 'meters',
+              },
+            },
+          }, function(err, results) {
+            should.not.exist(err);
+            should.exist(results);
```
Comment (superkhau) [inline test/mongodb.test.js]: please test if the values in results are what is expected

File: test/mongodb.test.js [hunk 5]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+            done();
+          });
+        });
+      });
+    });
+
+    it('find should be able to query by location via near with minDistance set', function(done) {
+      var coords = { lat: 30.274085, lng: 120.15507000000002 };
+      geoDb.autoupdate(function(err) {
+        var createPosts = function(far) {
```
Comment (superkhau) [inline test/mongodb.test.js]: ditto on shared function

File: test/mongodb.test.js [hunk 6]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+          };
+        };
+
+        async.parallel([
+          createPosts(false),
+          createPosts(false),
+          createPosts(false),
+          createPosts(true),
+        ], function(err) {
+          should.not.exist(err);
```
Comment (superkhau) [inline test/mongodb.test.js]: ditto

File: test/mongodb.test.js (hunk 4 above)
Comment (superkhau) [inline test/mongodb.test.js]: ditto

File: test/mongodb.test.js [hunk 7]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+            PostWithLocation.find({
+              where: {
+                location: {
+                  near: new GeoPoint(coords),
+                  maxDistance: distance,
+                  unit: unit,
+                },
+              },
+            }, function(err, results) {
+              should.not.exist(err);
```
Comment (superkhau) [inline test/mongodb.test.js]: ditto

File: test/mongodb.test.js [hunk 8]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+        };
+
+
+        async.parallel([
+          createPosts(false),
+          createPosts(false),
+          createPosts(false),
+          createPosts(true),
+        ], function(err) {
+          should.not.exist(err);
```
Comment (superkhau) [inline test/mongodb.test.js]: ditto

File: test/mongodb.test.js [hunk 9]
```diff
@@ -1438,6 +1438,184 @@ describe('mongodb connector', function() {
...
+          async.parallel([
+            queryLocation(10000, undefined, 10000, 3),
+            queryLocation(10, 'miles', 16000, 3),
+            queryLocation(10, 'kilometers', 10000, 3),
+            queryLocation(20000, 'feet', 6096, 3),
+            queryLocation(10000, 'radians', 10000, 3),
+            queryLocation(10000, 'degrees', 10000, 3),
+          ], function() {
```
Comment (superkhau) [inline test/mongodb.test.js]: Just pass in done directly.

`], done);`

**Comment by superkhau** [comment]
Some minor nitpicks, but looks mostly good. ;) Please ping me again after they've been addressed and let's get it merged.

**Comment by Tuxity** [comment]
Sure, updating right between 2 tasks at work ;)

File: test/mongodb.test.js [hunk 10]
```diff
@@ -1438,6 +1458,129 @@ describe('mongodb connector', function() {
...
+          if (err) return done(err);
+
+          async.parallel([
+            queryLocation(10000, undefined, 10000, 3),
+            queryLocation(10, 'miles', 16000, 3),
+            queryLocation(10, 'kilometers', 10000, 3),
+            queryLocation(20000, 'feet', 6096, 3),
+            queryLocation(10000, 'radians', 10000, 3),
+            queryLocation(10000, 'degrees', 10000, 3),
+          ], done;
```
Comment (superkhau) [inline test/mongodb.test.js]: done);

File: test/mongodb.test.js [hunk 11]
```diff
@@ -1438,6 +1458,129 @@ describe('mongodb connector', function() {
...
+              where: {
+                location: {
+                  near: new GeoPoint(coords),
+                  maxDistance: distance,
+                  unit: unit,
+                },
+              },
+            }, function(err, results) {
+              if (err) return done(err);
+              results.length.should.be.equal(numOfResult);
```
Comment (superkhau) [inline test/mongodb.test.js]: remove be

File: test/mongodb.test.js [hunk 12]
```diff
@@ -1438,6 +1458,129 @@ describe('mongodb connector', function() {
...
+            where: {
+              location: {
+                near: new GeoPoint(coords),
+                minDistance: 17000,
+                unit: 'meters',
+              },
+            },
+          }, function(err, results) {
+            if (err) return done(err);
+            should.exist(results);
```
Comment (superkhau) [inline test/mongodb.test.js]: remove as you already checked for errors on 1520

Comment (superkhau) [inline test/mongodb.test.js]: Actually you should leave this and verify the results are as expected. Meaning did you get the right values and number of objects back.

**Comment by Tuxity** [comment]
@superkhau It should be better now.
The unique doubt I have, is for `createPost` and my `createPosts` function. Should I merge them and relocate? where?

**Comment by superkhau** [comment]
@Tuxity Nah, it's fine. Small detail we can update later if needs be during a cleanup. Thanks for the contribution. LGTM. :+1:

**Comment by superkhau** [comment]
@slnode test please

**Comment by Tuxity** [comment]
Oups I made a mistake on tests <.< Fixing it

**Comment by Tuxity** [comment]
@superkhau It's updated, I had to put 4 instead of 1 because minDistance isn't fully working yet because of the missing implementation in loopback-datasource-juggler

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
test please

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by superkhau** [comment]
Closing in favour of #259

**Pull request closed without merging**